containerd.service

```diff
@@ -32,7 +32,7 @@ RestartSec=5
 # in the kernel. We recommend using cgroups to do container-local accounting.
 LimitNPROC=infinity
 LimitCORE=infinity
-LimitNOFILE=infinity
+LimitNOFILE=1048576
```

Comment:
@polarathene thanks for this suggestion, I tried to set the `LimitNOFILE=1024:524288` for both the `docker.service` and `containerd.service` using `override.conf` files in _/etc/systemd/system/docker.service.d_ and _/etc/systemd/system/containerd.service.d_ and noticed that **both** the hard limit and soft limit for `nofile` (within my container) were set to `524288`. I wonder if the soft limit from `LimitNOFILE` is ignored?

I know the `override.conf` files are "dropped-in":
```
● containerd.service - containerd container runtime
     Loaded: loaded (/usr/lib/systemd/system/containerd.service; enabled; preset: enabled)
    Drop-In: /etc/systemd/system/containerd.service.d
             └─override.conf
```
Output from within the container:
```
:~# docker run -it --network host --entrypoint bash container-name
:/# ulimit -Hn
524288
:/# ulimit -Sn
524288
```
Output on host:
```
:~# ulimit -Hn
524288
:~# ulimit -Sn
1024
```

Here is my output of `docker version`:

```
Client:
 Version:           23.0.1
 API version:       1.42
 Go version:        go1.19.7
 Git commit:        23.0.1
 Built:             unknown-buildtime
 OS/Arch:           linux/amd64
 Context:           default

Server:
 Engine:
  Version:          23.0.1
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.7
  Git commit:       buildroot
  Built:            
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.16
  GitCommit:        
 runc:
  Version:          1.1.4
  GitCommit:     
```

If I add the following to my _/etc/docker/daemon.json_ the ulimit hard and soft values for nofile is correct:

```
    "default-ulimits": {
        "nofile": {
            "Name": "nofile",
            "Hard": 524288,
            "Soft": 1024
        }
    }
```

Reply:
@dambrosio @polarathene it's a Golang issue, and will be fixed in the next patch release (the Go maintainers acknowledged a change they made in Go 1.19 was a regression, and a fix will be included in the next patch release);  see the discussion on this ticket, and the related backport tickets;

- https://github.com/golang/go/issues/46279
- https://github.com/golang/go/issues/59063
- https://github.com/golang/go/issues/59064

Reply:
@thaJeztah I am well aware and was involved in the discussions to get the issue addressed :sunglasses: 

---

@dambrosio my messages are a bit verbose sorry, but I did point this out in my review comment above:

> _whilst having the configured soft limit restored (pending backport to Go 1.19+) _

Once that is available, both `dockerd.service` and `containerd.service` should adjust `LimitNOFILE` as suggested, and the soft limit will be respected for containers (_without being an issue for either daemons needs as since Go 1.19_).

Reply:
The suggestion seems completely reasonable assuming the syntax works even on older versions of systemd. I think centos7 is v219?

Reply:
### Go updated with fix

Go 1.19.9 and Go 1.20.4 releases included backports for the `RLIMIT_NOFILE` syscall fix: https://github.com/golang/go/issues/59063

`contrainerd.service` and `docker.service` can now update their configs to move away from `LimitNOFILE=infinity` :tada:

Could the `LimitNOFILE=1024:524288` change request be applied, and queue this PR for review / merge? :grinning: 

---

### Confirmation

`containerd` 1.7.1 will include the needed Go update, and Docker Engine 23.0.6 includes the fix too. For Docker it seems to be resolved with 23.0.6, without the `containerd` release. Only need to override the current default `LimitNOFILE=infinity` in each service file.

Test values:
- **`containerd.service`:** `LimitNOFILE=2048:8192`
- **`docker.service`**: `LimitNOFILE=3072:4096`

```Dockerfile
FROM alpine
RUN echo "Soft: $(ulimit -Sn)" >> /limits.txt
RUN echo "Hard: $(ulimit -Hn)" >> /limits.txt
```

```console
# From v23 `docker build` uses LimitNOFILE from `docker.service`,
# Alternatively `DOCKER_BUILDKIT=0` will inherit `containerd.service` LimitNOFILE instead:
$ docker build --no-cache -t limits-test .
$ docker run --rm limits-test cat /limits.txt

Soft: 3072
Hard: 4096

# Silently ignores configured limit and falls back to inherited LimitNOFILE from service file:
# (If soft limit exceeds a hard limit specified/inherited, error will be output)
$ docker build --no-cache --ulimit 'nofile=4096:16384' -t limits-test .
$ docker run --rm limits-test cat /limits.txt

Soft: 3072
Hard: 4096

# So long as values don't exceed the hard limit inherited, adjustments apply:
$ docker build --no-cache --ulimit 'nofile=3080:3090' -t limits-test .
$ docker run --rm limits-test cat /limits.txt

Soft: 3080
Hard: 3090
```

```console
# LimitNOFILE from `containerd.service` applies to `docker run`:
$ docker run --rm alpine ash -c 'ulimit -Sn && ulimit -Hn'
2048
8192

# Same behaviour observed with `docker build` when hard limit exceeds inherited:
$ docker run --rm --ulimit 'nofile=4096:16384' alpine ash -c 'ulimit -Sn && ulimit -Hn'
2048
8192

# Same behaviour observed with `docker build` when valid override provided:
$ docker run --rm --ulimit 'nofile=1536:6144' alpine ash -c 'ulimit -Sn && ulimit -Hn'
1536
6144
```

This was tested on a Vultr VPS instance with Ubuntu 23.04 and [installing Docker via the docs](https://docs.docker.com/engine/install/ubuntu/):

- `docker-ce 23.0.6`
- `containerd containerd.io 1.6.21 3dce8eb055cbb6872793272b4f20ed16117344f8`
- `runc version: v1.1.7-0-g860f061`

<details>
<summary>docker info</summary>

```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.4
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.17.3
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 1
 Server Version: 23.0.6
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.2.0-20-generic
 Operating System: Ubuntu 23.04
 OSType: linux
 Architecture: x86_64
 CPUs: 1
 Total Memory: 945.4MiB
 Name: docker-limit
 ID: d94a5a6d-9d37-4244-bb85-6ca72526acdc
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

</details>